### DATA PT12 - Proyecto final - Grupo 2
#### Control de calidad de los datos del archivo Yelp reviews

In [1]:
import glob
import matplotlib.pyplot as plt
import numpy as np
import os
import pandas as pd
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

In [23]:
# Obtener la lista de archivos JSON en carpeta
#archivos_json = glob.glob(os.path.join("datos/yelp", "*.json"))

In [2]:
archivo = 'Yelp\\review.json'
chunks = pd.read_json(archivo, lines=True, chunksize=100000)  # Leer en bloques de 100,000 filas

# Guardar cada chunk en un archivo pickle
# Esto evita que la memoria se llene al procesar el JSON por partes.
for i, chunk in enumerate(chunks):
    chunk.to_pickle(f'Yelp\\review_part_{i}.pkl')  # Guarda en archivos más pequeños
    print(f'Chunk {i} guardado.')

Chunk 0 guardado.
Chunk 1 guardado.
Chunk 2 guardado.
Chunk 3 guardado.
Chunk 4 guardado.
Chunk 5 guardado.
Chunk 6 guardado.
Chunk 7 guardado.
Chunk 8 guardado.
Chunk 9 guardado.
Chunk 10 guardado.
Chunk 11 guardado.
Chunk 12 guardado.
Chunk 13 guardado.
Chunk 14 guardado.
Chunk 15 guardado.
Chunk 16 guardado.
Chunk 17 guardado.
Chunk 18 guardado.
Chunk 19 guardado.
Chunk 20 guardado.
Chunk 21 guardado.
Chunk 22 guardado.
Chunk 23 guardado.
Chunk 24 guardado.
Chunk 25 guardado.
Chunk 26 guardado.
Chunk 27 guardado.
Chunk 28 guardado.
Chunk 29 guardado.
Chunk 30 guardado.
Chunk 31 guardado.
Chunk 32 guardado.
Chunk 33 guardado.
Chunk 34 guardado.
Chunk 35 guardado.
Chunk 36 guardado.
Chunk 37 guardado.
Chunk 38 guardado.
Chunk 39 guardado.
Chunk 40 guardado.
Chunk 41 guardado.
Chunk 42 guardado.
Chunk 43 guardado.
Chunk 44 guardado.
Chunk 45 guardado.
Chunk 46 guardado.
Chunk 47 guardado.
Chunk 48 guardado.
Chunk 49 guardado.
Chunk 50 guardado.
Chunk 51 guardado.
Chunk 52 guardado.
Chu

In [4]:
import glob

# Obtener lista de archivos .pkl guardados
archivos_pkl = glob.glob('Yelp\\review_part_*.pkl')

# Leer y concatenar los DataFrames
df_reviews = pd.concat([pd.read_pickle(f) for f in archivos_pkl])

print(df_reviews.head())

                review_id                 user_id             business_id  \
0  KU_O5udG6zpxOg-VcAEodg  mh_-eMZ6K5RLWhZyISBhwA  XQfwVwDr-v0ZS3_CbbE5Xw   
1  BiTunyQ73aT9WBnpR9DZGw  OyoGAe7OKpv6SyGZT5g77Q  7ATYjTIgM3jUlt4UM3IypQ   
2  saUsX_uimxRlCVr67Z4Jig  8g_iMtfSiwikVnbP2etR0A  YjUWPpI6HXG530lwP-fb2A   
3  AqPFMleE6RsU23_auESxiA  _7bHUi9Uuf5__HHc_Q8guQ  kxX2SOes4o-D3ZQBkiMRfA   
4  Sx8TMOWLNuJBWer-0pcmoA  bcjbaE6dDog4jkNY91ncLQ  e4Vwtrqf-wpJfwesgvdgxQ   

   stars  useful  funny  cool  \
0      3       0      0     0   
1      5       1      0     1   
2      3       0      0     0   
3      5       1      0     1   
4      4       1      0     1   

                                                text                date  
0  If you decide to eat here, just be aware it is... 2018-07-07 22:09:11  
1  I've taken a lot of spin classes over the year... 2012-01-03 15:28:18  
2  Family diner. Had the buffet. Eclectic assortm... 2014-02-05 20:30:30  
3  Wow!  Yummy, different,  delicious.

In [5]:
print('text NaN', df_reviews['text'].isna().sum())
print('stars NaN', df_reviews['stars'].isna().sum())

text NaN 0
stars NaN 0


El dataframe reviews de Yelp no tiene valores nulos para la columna text (con los comentarios de las reseñas) ni para la columna stars (con la calificacion) 

In [6]:
# Convertir la columna a tipo datetime
# Vamos a asegurarnos de que la columna "date" esté en el formato correcto:
df_reviews['date'] = pd.to_datetime(df_reviews['date'], errors='coerce')

In [7]:
df_reviews.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6990280 entries, 0 to 999999
Data columns (total 9 columns):
 #   Column       Dtype         
---  ------       -----         
 0   review_id    object        
 1   user_id      object        
 2   business_id  object        
 3   stars        int64         
 4   useful       int64         
 5   funny        int64         
 6   cool         int64         
 7   text         object        
 8   date         datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(4)
memory usage: 533.3+ MB


In [8]:
df_reviews.head()

review_id                 user_id             business_id  \
0  KU_O5udG6zpxOg-VcAEodg  mh_-eMZ6K5RLWhZyISBhwA  XQfwVwDr-v0ZS3_CbbE5Xw   
1  BiTunyQ73aT9WBnpR9DZGw  OyoGAe7OKpv6SyGZT5g77Q  7ATYjTIgM3jUlt4UM3IypQ   
2  saUsX_uimxRlCVr67Z4Jig  8g_iMtfSiwikVnbP2etR0A  YjUWPpI6HXG530lwP-fb2A   
3  AqPFMleE6RsU23_auESxiA  _7bHUi9Uuf5__HHc_Q8guQ  kxX2SOes4o-D3ZQBkiMRfA   
4  Sx8TMOWLNuJBWer-0pcmoA  bcjbaE6dDog4jkNY91ncLQ  e4Vwtrqf-wpJfwesgvdgxQ   

   stars  useful  funny  cool  \
0      3       0      0     0   
1      5       1      0     1   
2      3       0      0     0   
3      5       1      0     1   
4      4       1      0     1   

                                                text                date  
0  If you decide to eat here, just be aware it is... 2018-07-07 22:09:11  
1  I've taken a lot of spin classes over the year... 2012-01-03 15:28:18  
2  Family diner. Had the buffet. Eclectic assortm... 2014-02-05 20:30:30  
3  Wow!  Yummy, different,  delicious.   Our favo... 2015-01-04 00:01:03  
4  Cute interior and owner (?) gave us tour of up... 2017-01-14 20:54:15

El dataframe tiene 6.990.280 reviews. La columna date esta en formato datetime.

In [10]:
# df_rev.to_parquet('/yelp/reviews.parquet' , engine='fastparquet')
# El motor pyarrow maneja mejor archivos grandes y evita problemas de compresión:
df_reviews.to_parquet('Yelp\\reviews.parquet', engine='pyarrow', compression='snappy')

In [11]:
df_dup = df_reviews[df_reviews.duplicated(subset=['business_id',  'date', 'review_id', 'user_id'], keep=False)]
df_dup = df_dup.sort_values([])

In [12]:
print(df_dup)

Empty DataFrame
Columns: [review_id, user_id, business_id, stars, useful, funny, cool, text, date]
Index: []


No hay filas duplicadas

In [13]:
df_reviews.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6990280 entries, 0 to 999999
Data columns (total 9 columns):
 #   Column       Dtype         
---  ------       -----         
 0   review_id    object        
 1   user_id      object        
 2   business_id  object        
 3   stars        int64         
 4   useful       int64         
 5   funny        int64         
 6   cool         int64         
 7   text         object        
 8   date         datetime64[ns]
dtypes: datetime64[ns](1), int64(4), object(4)
memory usage: 533.3+ MB


In [14]:
# leo el archivo de las pizzerias mt_pizza_NJNY para hacer join con este
# How to read a Parquet file into Pandas DataFrame?
business = pd.read_parquet('Yelp\\business.parquet' , engine='fastparquet')

In [15]:
business.info()

<class 'pandas.core.frame.DataFrame'>
Index: 635 entries, 29 to 150172
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   address       635 non-null    object 
 1   business_id   635 non-null    object 
 2   city          635 non-null    object 
 3   latitude      635 non-null    float64
 4   longitude     635 non-null    float64
 5   name          635 non-null    object 
 6   postal_code   635 non-null    object 
 7   review_count  635 non-null    int64  
 8   stars         635 non-null    float64
 9   state         635 non-null    object 
dtypes: float64(3), int64(1), object(6)
memory usage: 54.6+ KB


In [16]:
business_pizza = pd.merge(business, df_reviews, how='left', on='business_id') # tengo que hacer left join

In [17]:
business_pizza.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 41946 entries, 0 to 41945
Data columns (total 18 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   address       41946 non-null  object        
 1   business_id   41946 non-null  object        
 2   city          41946 non-null  object        
 3   latitude      41946 non-null  float64       
 4   longitude     41946 non-null  float64       
 5   name          41946 non-null  object        
 6   postal_code   41946 non-null  object        
 7   review_count  41946 non-null  int64         
 8   stars_x       41946 non-null  float64       
 9   state         41946 non-null  object        
 10  review_id     41946 non-null  object        
 11  user_id       41946 non-null  object        
 12  stars_y       41946 non-null  int64         
 13  useful        41946 non-null  int64         
 14  funny         41946 non-null  int64         
 15  cool          41946 non-null  int64 

Hay 41946 reseñas de Yelp para las pizzerias del estado

In [18]:
# Modifico los nombres de las columnas que estan duplicados
business_pizza.rename(columns={'stars_x': 'stars_business', 'stars_y': 'stars_by_user'}, inplace=True)

In [19]:
business_pizza.head()

address             business_id         city   latitude  \
0  3001 Highway 31 W  sqSqqLy0sN8n2IZrAbzidQ  White House  36.464747   
1  3001 Highway 31 W  sqSqqLy0sN8n2IZrAbzidQ  White House  36.464747   
2  3001 Highway 31 W  sqSqqLy0sN8n2IZrAbzidQ  White House  36.464747   
3  3001 Highway 31 W  sqSqqLy0sN8n2IZrAbzidQ  White House  36.464747   
4  3001 Highway 31 W  sqSqqLy0sN8n2IZrAbzidQ  White House  36.464747   

   longitude            name postal_code  review_count  stars_business state  \
0 -86.659187  Domino's Pizza       37188             8             3.5    CA   
1 -86.659187  Domino's Pizza       37188             8             3.5    CA   
2 -86.659187  Domino's Pizza       37188             8             3.5    CA   
3 -86.659187  Domino's Pizza       37188             8             3.5    CA   
4 -86.659187  Domino's Pizza       37188             8             3.5    CA   

                review_id                 user_id  stars_by_user  useful  \
0  Nd9IlLVwTdG0JEFYnA4cJw  _GZUfylSfVj2gNOkiqsSoQ              1       0   
1  gn8_q0-x36PvNli6Rc_Upw  -gwn8fgTfq18P_nfA-8dUQ              5       0   
2  5ekFz7qmWJNxT6uvM3HBBg  oRQRl1Dw6OZrVwi4y0K9Kg              4       0   
3  wrQ7t43gqQ4kIcWyuZYKpw  5BpQb0e4i9gw0HUjOpYUYA              5       1   
4  MVEKSpJkjogSNRSbJ-0r7w  IN9v5sjMnlDfq2pid3pwBw              2       1   

   funny  cool                                               text  \
0      0     0  Undercook and overpaid! Got a pizza and it was...   
1      0     0  We always have great service here. Only once i...   
2      0     0  Dominoes is getting better lately about gettin...   
3      0     0  Great pizzas and great service.  Love the orde...   
4      0     0  Ordered online, quoted 26-36 min, delivered al...   

                 date  
0 2018-08-26 14:44:11  
1 2015-08-02 02:36:16  
2 2017-04-04 01:00:51  
3 2014-04-13 00:15:58  
4 2016-10-22 01:03:48

In [20]:
print(business_pizza.iloc[[0]])

             address             business_id         city   latitude  \
0  3001 Highway 31 W  sqSqqLy0sN8n2IZrAbzidQ  White House  36.464747   

   longitude            name postal_code  review_count  stars_business state  \
0 -86.659187  Domino's Pizza       37188             8             3.5    CA   

                review_id                 user_id  stars_by_user  useful  \
0  Nd9IlLVwTdG0JEFYnA4cJw  _GZUfylSfVj2gNOkiqsSoQ              1       0   

   funny  cool                                               text  \
0      0     0  Undercook and overpaid! Got a pizza and it was...   

                 date  
0 2018-08-26 14:44:11  


In [21]:
# guardo df_pizza_NJNY_rev
business_pizza.to_parquet('Yelp\\business_pizza_rev.parquet' , engine='fastparquet')